tests/bench_xtimer_load/main.c

```diff
@@ -0,0 +1,213 @@
+ *
+ * @}
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+
+#include "xtimer.h"
+#include "thread.h"
+#include "msg.h"
+#include "log.h"
+
+/* We generate some context switching and IPC traffic by using multiple threads
+ * and generate some xtimer load by scheduling several messages to be called at
+ * different times. TEST_HZ is the frequency of messages being sent from the
+ * main thread to the worker, all other message frequencies are derived from
+ * TEST_HZ.
+ * TEST_MSG_RX_USLEEP is a tiny sleep inside the message reception thread to
+ * cause extra context switches.
+ */
+#ifndef TEST_HZ
+#define TEST_HZ             (16LU)
+#endif
+#define TEST_INTERVAL       (1000000LU / TEST_HZ)
+#define TEST_MSG_RX_USLEEP  (200LU)
+#define TEST_MSG_QUEUE_SIZE (4U)
+#define TEST_TIME           (10U)
+
+char slacker_stack1[THREAD_STACKSIZE_DEFAULT];
+char slacker_stack2[THREAD_STACKSIZE_DEFAULT];
+char worker_stack[THREAD_STACKSIZE_MAIN];
+
+struct timer_msg {
+    xtimer_t timer;
+    uint32_t interval;
+    msg_t msg;
+};
+
+struct timer_msg msg_a = { .interval = (TEST_INTERVAL / 2) };
+struct timer_msg msg_b = { .interval = (TEST_INTERVAL / 3) };
+struct timer_msg msg_c = { .interval = (TEST_INTERVAL * 5) };
+struct timer_msg msg_d = { .interval = (TEST_INTERVAL * 2) };
+
+/* This thread is only here to give the kernel some extra load */
+void *slacker_thread(void *arg)
+{
+    (void)arg;
+    timex_t now;
+
+    LOG_DEBUG("run thread %" PRIkernel_pid "\n", thread_getpid());
+
+    /* we need a queue if a 2nd message arrives while the first is processed */
+    msg_t msgq[TEST_MSG_QUEUE_SIZE];
+    msg_init_queue(msgq, TEST_MSG_QUEUE_SIZE);
+
+    while (1) {
+        msg_t m;
+        msg_receive(&m);
+        struct timer_msg *tmsg = m.content.ptr;
+        xtimer_now_timex(&now);
+        xtimer_usleep(TEST_MSG_RX_USLEEP);
+
+        tmsg->msg.type = 12345;
+        tmsg->msg.content.ptr = tmsg;
+        xtimer_set_msg(&tmsg->timer, tmsg->interval, &tmsg->msg,
+                       thread_getpid());
+    }
+}
+
+static volatile int32_t _min_drift, _max_drift, _min_jitter, _max_jitter;
+static volatile int32_t _final_drift;
+static volatile uint32_t _total_jitter, _samples;
+
+/* This thread will print the drift to stdout once per second */
+void *worker_thread(void *arg)
+{
+    (void)arg;
+
+    /* Calculate interval based on possible precision when 'XTIMER_SHIFT > 0',
+     * to apply precision loss to expected interval length.
+     * test_interval != TEST_INTERVAL */
+    uint32_t test_interval =
+        xtimer_usec_from_ticks(xtimer_ticks_from_usec(TEST_INTERVAL));
+    uint32_t start = 0;
+    uint32_t last = 0;
+    uint32_t loop_counter = 0;
+
+    LOG_DEBUG("run thread %" PRIkernel_pid "\n", thread_getpid());
+
+    while (1) {
+        msg_t m;
+        msg_receive(&m);
+
+        xtimer_ticks32_t ticks = xtimer_now();
+        uint32_t now = xtimer_usec_from_ticks(ticks);
```

Comment:
Is there a reason for not using `xtimer_now_usec()` here?

Reply:
nope, fixed